ext/openssl/ossl.c

```diff
@@ -604,6 +604,30 @@ static void Init_ossl_locks(void)
 }
 #endif /* !HAVE_OPENSSL_110_THREADING_API */
 
+/*
+ * call-seq:
+ *   OpenSSL.memcmp?("aaa", "aaa")   => true
+ *   OpenSSL.memcmp?("aaa", "bbb")   => false
+ *
+ * Constant time memory comparison
+ */
+static VALUE
+ossl_crypto_memcmp(VALUE dummy, VALUE str1, VALUE str2)
+{
+    const unsigned char *p1 = (const unsigned char *)StringValuePtr(str1);
+    const unsigned char *p2 = (const unsigned char *)StringValuePtr(str2);
+    long len1 = RSTRING_LEN(str1);
+    long len2 = RSTRING_LEN(str2);
+
+    if(len1 != len2)
```

Comment:
For a crytographic function that's trying to be timing invariant, is this acceptable?

Reply:
We have to do something to ensure the strings are of equal length, otherwise the `refute OpenSSL.memcmp?("aaa", "aaaa")` test fails since [CRYPTO_memcmp](https://www.openssl.org/docs/manmaster/man3/CRYPTO_memcmp.html) compares up to `len1` bytes.

Rack, Active Support up to 5.1 (https://github.com/rails/rails/pull/24510), .NET Core and the snippets in https://codahale.com/a-lesson-in-timing-attacks/ do the same thing and exit early. I think it's not terrible, calling it out in the docs would be the least we can do.

Active Support 5.2+ replaced the default behaviour by always hashing with SHA-256 first and raising on length inequality in a separate method. I can do something similar if you prefer.

Reply:
@picatz do you have any input on this?

Here is an example which shows that such an implementation leaks the string length:

```ruby
#!/usr/bin/env ruby

require 'benchmark'

def compare(a, b)
	return false unless a.bytesize == b.bytesize

	l = a.unpack "C#{a.bytesize}"

	res = 0
	b.each_byte { |byte| res |= byte ^ l.shift }
	res == 0
end


def find_length_of_secret(secret)
	samples = 100.times.collect do |i|
		test = "\0" * i
		
		[i, Benchmark.realtime do
			1000.times do
				compare(test, secret)
			end
		end]
	end
	
	best_guess = samples.sort_by{|sample| sample.last}.last
	
	return best_guess.first
end

pp find_length_of_secret("foobar" * 2)
```

The example you give from ActiveSupport throws an exception when strings are not the same length. Provided we clearly mark the arguments (user input) and (secret), we can run the algorithm up to the length of secret in every case, that will minimise leaking any details about the length of the secret string... the entire runtime of the algorithm might be used to deduce the string length. For example:

```ruby
compare(user_input, "twenty characters!!!") -> 20ns
compare(user_input, secret) -> 40ns
```

Maybe user can figure out secret is 40 characters long.

But, given how much noise and other overheads we expect, maybe it's not a concern. Even though I can think of several ways to make this harder, I'm not sure it's feasible to fix - e.g. adding a small random amount of noise to the function timing (`sleep rand`).

Reply:
Maybe throwing an exception if the string lengths are different isn't such a stupid idea?

Reply:
After thinking about this a bit more, I think it's the most safe option to pick yeah. The OpenSSL variant might be harder to attack than the Ruby one due to the lower speed variance as seen in the benchmark. Or the length of the secret might not be much of a secret in itself (e.g. 32 bytes/SHA-256). But then again you might never know how people use this API and how that can be abused 🤷‍♂

I'll work on this change and some additional documentation tomorrow.

Reply:
Maybe do `CRYPTO_memcmp('a', 'b', 1)` (or another example that always returns false if the lenghts are not the same?

Reply:
In ruby psuedo code you could do something like this:
```
def ossl_crypto_memcmp(str1, str2)
  len1 = str1.length
  len2 = str2.length
  # always generate incorrect answers even when str1 and str2 are the same length so it doesn't effect the time.
  incorrect1 = 'a' * len1
  incorrect2 = 'b' * len1
  if len1 != len2
    CRYPTO_memcmp(incorrect1 , incorrect2, len1)
  else
    CRYPTO_memcmp(str1, str2, len1)
  end
end
```
But reading the documentation for CRYPTO_memcmp it can still leak the length of the other string. Using this would require users to always make sure the strings to compare are always the same length. 
> The CRYPTO_memcmp function compares the len bytes pointed to by a and b for equality. It takes an amount of time dependent on len, but independent of the contents of the memory regions pointed to by a and b.

Reply:
Here's a good discussion about the keeping the length secret https://security.stackexchange.com/a/92238